web/src/components/product/ProductRegistrationPage.tsx

```diff
@@ -174,9 +204,112 @@ const HostnameAlert = () => {
+  const [regCode, setRegCode] = useState("");
+  const [error, setError] = useState<string | null>(null);
+  const [loading, setLoading] = useState(false);
+
+  const onRegisterError = ({ response }) => {
+    setError(response.data.message);
+  };
+
+  const registered = registeredExtensions.find(
+    (e) => e.id === extension.id && (e.version === extension.version || e.version === null),
+  );
+
+  const submit = async (e: React.SyntheticEvent) => {
+    e.preventDefault();
+    setLoading(true);
+
+    const data: RegisteredAddonInfo = {
+      id: extension.id,
+      registrationCode: regCode,
+      // omit the version if only one version of the extension exists
+      version: unique ? null : extension.version,
+    };
+
+    // @ts-expect-error
+    registerAddon(data, { onError: onRegisterError, onSettled: () => setLoading(false) });
+  };
+
+  return (
+    <div>
+      {/* remove the "(BETA)" suffix, we display a Beta label instead */}
+      <Title headingLevel="h4">
+        {extension.label.replace(/\s*\(beta\)$/i, "")}{" "}
+        {extension.release === "beta" && (
+          <Label color="blue" isCompact>
+            {_("Beta")}
+          </Label>
+        )}
+        {extension.recommended && (
+          <Label color="orange" isCompact>
+            {_("Recommended")}
+          </Label>
+        )}
+      </Title>
+      <p>&nbsp;</p>
+      <p>{extension.description}</p>
+      <p>&nbsp;</p>
+      <p>
+        {registered ? (
+          <RegisteredExtensionSection extension={registered} />
+        ) : extension.available ? (
+          <Form id={`register-form-${extension.id}-${extension.version}`} onSubmit={submit}>
+            {error && <Alert variant="warning" isInline title={error} />}
+            {!extension.free && (
+              <FormGroup label={KEY_LABEL}>
+                <RegistrationCodeInput
+                  id={`reg-code-${extension.id}-${extension.version}`}
+                  value={regCode}
+                  onChange={(_, v) => setRegCode(v)}
+                />
+              </FormGroup>
+            )}
+
+            <ActionGroup>
+              <Button variant="primary" type="submit" isInline isLoading={loading}>
+                {_("Register")}
+              </Button>
+            </ActionGroup>
+          </Form>
+        ) : (
+          <Alert title={_("Not available")} variant="warning">
+            {_(
+              "This extension is not available on the server. Please ask the server administrator to mirror the extension.",
+            )}
+          </Alert>
+        )}
+      </p>
+    </div>
+  );
+};
+
+const Extensions = () => {
+  const extensions = useAddons();
+
+  // count the extension versions
+  const counts = {};
+  extensions.forEach((e) => (counts[e.id] = counts[e.id] ? counts[e.id] + 1 : 1));
+
+  const extensionComponents = extensions.map((ext, index) => (
+    <Extension key={`extension-${index}`} extension={ext} unique={counts[ext.id] === 1} />
```

Comment:
Hard to understand the count thingy. Isn't there another way to know if the extension is unique?

What is more, 

`(counts[e.id] = counts[e.id] ? counts[e.id] + 1 : 1)`

Looks a not valid expression to me. I mean, it will go always for the true branch because there is no a comparison but an assignation.

Although since you have tested the code I suspect I'm missing something. Do you mind explaining this better?

Reply:
```
(counts[e.id] = counts[e.id] ? counts[e.id] + 1 : 1)
                ^^^^^^^^^^^^
```

There is an implicit comparison, the `counts[e.id]` above is evaluated as boolean. If the key does not exist it returns `undefined` which in boolean context is evaluated as `false`. If the key exists the returned number is evaluated as `true`.

Reply:
Still quite tricky, but if there is no other option just go ahead.

Reply:
I have changed that.